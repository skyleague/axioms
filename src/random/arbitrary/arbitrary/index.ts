export { Arbitrary, TypeOfArbitrary, TypeOfArbitraries, interleaveTree, interleave, interleaveList } from './arbitrary'
