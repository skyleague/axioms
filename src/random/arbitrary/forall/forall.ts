import { isJust } from '../../../guard'
import { replicate } from '../../../iterator/replicate'
import { Nothing } from '../../../type/maybe'
import type { Maybe } from '../../../type/maybe'
import { xoroshiro128plus } from '../../rng/xoroshiro128plus'
import type { Arbitrary } from '../arbitrary'
import type { ArbitraryContext } from '../context'
import { falsify, FalsifiedError } from '../falsify'
import { InfeasibleTree } from '../shrink'

export interface ForallOptions<T> {
    tests: number
    shrinks: number
    period: number
    maxSkips: number
    seed?: bigint
    counterExample?: T
}

export function forAll<T>(
    arbitrary: Arbitrary<T>,
    predicate: (x: T, context: ArbitraryContext) => boolean | void,
    {
        tests = 100,
        shrinks = 200,
        maxSkips = 100,
        seed = BigInt(new Date().getTime()),
        period = 13,
        counterExample,
    }: Partial<ForallOptions<T>> = {}
): void {
    const context: ArbitraryContext = {
        rng: xoroshiro128plus(BigInt(seed)),
        bias: undefined,
    }
    function safePredicate(x: T): [boolean, Maybe<Error>] {
        let holds: boolean | undefined | void
        let error: Maybe<Error> = Nothing
        try {
            holds = predicate(x, context)
        } catch (e: unknown) {
            holds = false
            error = e as Error
        }
        // void returned a proper result, which means the predicate held
        return [typeof holds === 'boolean' ? holds : true, error]
    }
    const maybeCounterExample = falsify<T>({
        predicate: safePredicate,
        values: (ctx = { skips: 0 }) =>
            replicate((i) => {
                while (ctx.skips < maxSkips) {
                    try {
                        const value = arbitrary.value({
                            ...context,
                            bias: context.rng.sample() < 1 / (1 + Math.log(i + 1)) ? context.rng.sample() : undefined,
                            // bias: undefined,
                        })

                        if (i > 0 && i % period === 0) {
                            context.rng.jump()
                        }
                        return value
                    } catch (e) {
                        if (e instanceof InfeasibleTree) {
                            ctx.skips++
                        } else {
                            throw e
                        }
                    }
                }
                throw new InfeasibleTree()
            }, tests),
        maxDepth: shrinks,
        counterExample,
    })

    if (isJust(maybeCounterExample)) {
        throw new FalsifiedError(maybeCounterExample, { seed })
    }
}
