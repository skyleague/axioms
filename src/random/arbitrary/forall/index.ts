export { ForallOptions, forAll } from './forall'
