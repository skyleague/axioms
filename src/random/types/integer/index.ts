export { IntegerConstraints, integer } from './integer'
