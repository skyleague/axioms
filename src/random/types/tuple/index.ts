export { tuple } from './tuple.js'
