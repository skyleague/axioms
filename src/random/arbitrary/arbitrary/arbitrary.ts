import { mapApplicativeTree, tree } from '../../../algorithm/tree'
import type { Tree } from '../../../algorithm/tree'
import { collect } from '../../../array/collect'
import { curryTuple } from '../../../function/tuple'
import { next } from '../../../generator'
import { isRight } from '../../../guard'
import { concat, map } from '../../../iterator'
import type { Traversable } from '../../../type'
import { toTraverser } from '../../../type'
import type { ArbitraryContext } from '../context'
import { shrinkX, shrinkOne } from '../shrink'

export interface Arbitrary<T> {
    value: (context: ArbitraryContext) => Tree<T>
}

export type TypeOfArbitrary<T extends Arbitrary<unknown>> = ReturnType<T['value']>['value']
export type TypeOfArbitraries<T extends Arbitrary<unknown>[]> = ReturnType<[...T][number]['value']>['value']

export function interleaveTree<T, U>(r: Tree<T>, l: Tree<(x: T) => U>): Tree<U> {
    const { value: f, children: ls } = l
    const { value: x, children: rs } = r
    return {
        value: f(x),
        children: concat(
            map(ls, (lp) => interleaveTree(r, lp)),
            map(rs, (rp) => interleaveTree(rp, l))
        ),
    }
}

// @todo: optimize this
export function interleave<U extends Tree<unknown>[]>(
    ...xs: [...U]
): Tree<{ [K in keyof U]: U[K] extends { value: infer Value } ? Value : never }> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let init: any = mapApplicativeTree(curryTuple(xs.length), xs[0])
    for (let i = 1; i < xs.length; ++i) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
        init = interleaveTree(xs[i], init)
    }
    return init as Tree<{ [K in keyof U]: U[K] extends { value: infer Value } ? Value : never }>
}

export function shrinkAll<T>(xs: Tree<T>[]): Tree<T[]>[] {
    const children = xs.map((x) => {
        const it = toTraverser(x.children)
        const n = next(it)
        return isRight(n) ? ([true, n.right] as const) : ([false, x] as const)
    })
    return children.every(([shrunk]) => shrunk) ? [interleaveList(map(children, ([, c]) => c))] : []
}

export function interleaveList<T>(xs: Traversable<Tree<T>>, options: { minLength?: number } = {}): Tree<T[]> {
    const { minLength = 0 } = options
    const axs = collect(xs)
    return tree(
        axs.map((x) => x.value),
        concat(
            // half first to dissect
            ...[Math.floor(axs.length * 0.5) > minLength ? [interleaveList(shrinkX(axs, 0.5), options)] : []],
            ...map(shrinkOne(axs), ([as, b, cs]) => map(b.children, (c) => interleaveList(concat(as, [c], cs), options))),
            ...[axs.length > minLength ? map(shrinkOne(axs), ([as, , cs]) => interleaveList(concat(as, cs), options)) : []]
            // lazy(() => shrinkAll(axs))
        )
    )
}
