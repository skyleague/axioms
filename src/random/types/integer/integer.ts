import { expandTree, tree } from '../../../algorithm'
import type { Tree } from '../../../algorithm'
import type { RelaxedPartial } from '../../../type'
import { makeIntegrated, towards } from '../../arbitrary'
import type { ArbitraryContext, BiasedArbitraryContext, Integrated } from '../../arbitrary'
import { weightedChoice } from '../choice'

export interface IntegerConstraints {
    min: number
    max: number
}

const nearZeroBias = weightedChoice([
    [2, ({ logMin, logMax }: IntegerConstraints & { logMin: number; logMax: number }) => ({ min: -logMin, max: logMax })],
    [1, ({ logMax, max }: IntegerConstraints & { logMin: number; logMax: number }) => ({ min: max - logMax, max })],
    [1, ({ logMin, min }: IntegerConstraints & { logMin: number; logMax: number }) => ({ min, max: min + logMin })],
])

export function integerLogLike(v: number): number {
    return Math.floor(Math.log(v))
}

export function sampleInteger({ min, max }: IntegerConstraints, { rng }: ArbitraryContext): number {
    return Math.floor(rng.sample() * (max - min) + min)
}

export function biasInteger({ min, max }: IntegerConstraints, { rng, bias }: BiasedArbitraryContext): IntegerConstraints {
    if (min === max) {
        return { min, max }
    } else if (min < 0 && max > 0) {
        // min < 0 && max > 0
        const logMin = integerLogLike(-min) * bias
        const logMax = integerLogLike(max) * bias

        return nearZeroBias(rng.sample())({ min, max, logMin, logMax })
    }
    // // Either min < 0 && max <= 0
    // // Or min >= 0, so max >= 0
    const length = (max - min) * bias

    const choices = weightedChoice([
        [1, { min, max: Math.floor(min + length) }],
        [1, { min: Math.floor(max - length), max }],
    ])

    // const logGap = integerLogLike((max - min) as any) // max-min !== 0
    // const arbCloseToMin = new Ctor(min, max, min, (min as any) + logGap) // close to min
    // const arbCloseToMax = new Ctor(min, max, (max - logGap) as any, max) // close to max
    // return min < 0
    //     ? new BiasedNumericArbitrary(arbCloseToMax, arbCloseToMin) // max is closer to zero
    //     : new BiasedNumericArbitrary(arbCloseToMin, arbCloseToMax) // min is closer to zero
    return choices(rng.sample())
}

export function shrinkInteger({ min, max }: IntegerConstraints, x: number): Tree<number> {
    const destination = min <= 0 && max >= 0 ? 0 : min < 0 ? max : min
    return expandTree((v) => towards(v, destination), tree(x, [tree(destination)]))
}

export function integer(constraints: RelaxedPartial<IntegerConstraints> = {}): Integrated<IntegerConstraints, number> {
    const { min = -Math.pow(2, 31), max = Math.pow(2, 31) } = constraints

    return makeIntegrated({
        sample: sampleInteger,
        biased: biasInteger,
        shrink: shrinkInteger,
        constraints: {
            min,
            max,
        },
    })
}
