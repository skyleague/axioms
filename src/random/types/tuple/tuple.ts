import type { Arbitrary } from '../../arbitrary/arbitrary'
import { interleave } from '../../arbitrary/arbitrary'
import type { ArbitraryContext } from '../../arbitrary/context'
import type { Dependent } from '../../arbitrary/dependent'
import { makeDependent } from '../../arbitrary/dependent'

export function tuple<T extends Arbitrary<unknown>[]>(
    ...xs: [...T]
): Dependent<{ [K in keyof T]: T[K] extends { value(context: ArbitraryContext): { value: infer Value } } ? Value : never }> {
    return makeDependent((context) => interleave(...xs.map((x) => x.value(context)))) as Dependent<{
        [K in keyof T]: T[K] extends { value(context: ArbitraryContext): { value: infer Value } } ? Value : never
    }>
}
