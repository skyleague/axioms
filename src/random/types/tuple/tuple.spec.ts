import { tuple } from './tuple'

import { forAll } from '../../arbitrary/forall'
import { integer } from '../integer'

test('counter example - equal', () => {
    expect(() => forAll(tuple(integer(), integer()), ([a, b]) => a !== b, { seed: 42n })).toThrowErrorMatchingInlineSnapshot(`
        "Counter example found after 43 tests (seed: 42n)
        Shrunk 0 time(s)
        Counter example:

        [ 2147483645, 2147483645 ]"
    `)
})
