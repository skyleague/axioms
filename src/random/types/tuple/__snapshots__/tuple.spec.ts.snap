// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`show small tree - integers 1`] = `
"└─ 15,17
    ├─ 0,0
    ├─ 0,17
    |   ├─ 0,0
    |   ├─ 0,9
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 0,13
    |   |   └─...
    |   |   └─...
    |   ├─ 0,15
    |   |   └─...
    |   └─ 0,16
    ├─ 8,17
    |   ├─ 4,0
    |   |   └─...
    |   |   └─...
    |   ├─ 4,17
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 6,17
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 7,17
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 8,0
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 8,9
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 8,13
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 8,15
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 8,16
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   └─ 6,9
    |       └─...
    |       └─...
    |       └─...
    |       └─...
    |       └─...
    ├─ 12,17
    |   ├─ 10,0
    |   |   └─...
    |   ├─ 10,17
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 11,17
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 12,0
    |   |   └─...
    |   |   └─...
    |   ├─ 12,9
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 12,13
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 12,15
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 12,16
    |   |   └─...
    |   |   └─...
    |   └─ 11,9
    |       └─...
    |       └─...
    |       └─...
    ├─ 14,17
    |   ├─ 13,0
    |   ├─ 13,17
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 14,0
    |   |   └─...
    |   ├─ 14,9
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 14,13
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 14,15
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   └─ 14,16
    |       └─...
    ├─ 15,0
    |   ├─ 0,0
    |   ├─ 8,0
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 12,0
    |   |   └─...
    |   |   └─...
    |   └─ 14,0
    |       └─...
    ├─ 15,9
    |   ├─ 0,5
    |   |   └─...
    |   |   └─...
    |   ├─ 0,9
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 8,9
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 12,9
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 14,9
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 15,5
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 15,7
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 15,8
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   └─ 8,7
    |       └─...
    |       └─...
    |       └─...
    |       └─...
    |       └─...
    ├─ 15,13
    |   ├─ 0,11
    |   |   └─...
    |   ├─ 0,13
    |   |   └─...
    |   |   └─...
    |   ├─ 8,13
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 12,13
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 14,13
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 15,11
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 15,12
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   └─ 8,12
    |       └─...
    |       └─...
    |       └─...
    ├─ 15,15
    |   ├─ 0,14
    |   ├─ 0,15
    |   |   └─...
    |   ├─ 8,15
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 12,15
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 14,15
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   └─ 15,14
    |       └─...
    |       └─...
    |       └─...
    |       └─...
    ├─ 15,16
    |   ├─ 0,16
    |   ├─ 8,16
    |   |   └─...
    |   |   └─...
    |   |   └─...
    |   ├─ 12,16
    |   |   └─...
    |   |   └─...
    |   └─ 14,16
    |       └─...
    └─ 8,9
        ├─ 4,5
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        ├─ 4,9
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        ├─ 6,9
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        ├─ 7,9
        |   └─...
        |   └─...
        |   └─...
        ├─ 8,5
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        ├─ 8,7
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        |   └─...
        ├─ 8,8
        |   └─...
        |   └─...
        |   └─...
        └─ 6,7
            └─...
            └─...
            └─..."
`;
